proposals/unions.md

```diff
@@ -0,0 +1,322 @@
+
+## Detailed design
+[design]: #detailed-design
+
+### Union interfaces
+
+The following interfaces are recognized by the language in its implementation of union features.
+
+#### Union access interface
+
+The `IUnion` interface marks a type as a union type at compile time and provides a way to access union contents at runtime.
+
+```csharp
+public interface IUnion
+{
+    // The value of the union or null
+    object? Value { get; }
+}
+```
+As we'll see later:
+* The `IUnion` interface is all that is necessary for a type to be considered a union by the language.  
+* The `Value` property of the interface is targeted by the compiler to implement pattern matching.
+* Union declarations are generated by the compiler to implement this interface.
+
+Example use:
+
+```csharp
+if (value is IUnion union && union.Value is null) { ... }
+```
+
+#### Union creation interface
+
+The `IUnion<TUnion>` interface provides a means to create union instances at runtime.
+
+```csharp
+public interface IUnion<TUnion> : IUnion where TUnion : IUnion<TUnion>
+{
+    // Creates a union from a value
+    static abstract bool TryCreate(object? value, [NotNullWhen(true)] out TUnion union);
+}
+```
+
+* Union declarations are generated by the compiler to implement this interface.
+
+Example use:
+
+```csharp
+TUnion ReadUnion<TUnion>() where TUnion : IUnion<TUnion>
+{
+    object val = ReadValue();
+
+    // wrap the value in the union
+    if (TUnion.TryCreate(val, out var union)) 
+        return union;
+
+    throw ...;
+}
+```
+
+### Union types
+
+Any class or struct type that implements the `IUnion` interface is considered a *union type*. For each public constructor with exactly one parameter, the type of that parameter is considered a *case type* of the union type.
+
+The contents of a union value can be accessed through the `IUnion.Value` property. The language assumes that `Value` only ever contains a value of one of the case types, or null.
+
+Example of a union type:
+
+```csharp
+public struct MyUnion : IUnion
+{
+    public MyUnion(Case1 value) { this.Value = value; }
+    public MyUnion(Case2 value) { this.Value = value; }
+
+    // implements IUnion.Value
+    public object? Value { get; }
+}
+```
+
+`MyUnion` has the two case types `Case1` and `Case2`.
+
+### Union patterns
+
+Union types always exhibit the *basic union pattern*, but may additionally implement the *non-boxing union access pattern*.
+
+- *The basic union pattern* consists of the `IUnion.Value` property and the public constructors that give rise to the case types of the union type. All of the union behaviors can be implemented in terms of this pattern.
+- *The non-boxing union access pattern* adds a `public bool HasValue { get; }` property, as well as a `public bool TryGetValue(out T value)` method for each case type `T`. This allows the union pattern matching behavior to be implemented without boxing.
+
+Example of a union type implementing the non-boxing access pattern:
```

Comment:
perhaps have a comment somehwere explicitly calling out that the internal implementation of this example is entirely free to change, using any internal representation the user chooses.  

in other words, make it slightly clearer that the non-boxing pattern doesn't  mean you have to store it as an enum + fields for each case.